binary_tree.c

```diff
@@ -0,0 +1,410 @@
+
+    if(self->l_node) {
+        new_node->l_node=self->l_node;
+    }
+    self->l_node=new_node;
+}
+
+void tree_insert_right(struct Node* self, int value) {
+    struct Node* new_node = tree_create_node(value);
+
+    if(self->r_node) {
+        new_node->r_node=self->r_node;
+    }
+
+    self->r_node=new_node;
+}
+
+struct Node* tree_bfs_search(struct Node* self, int value) {
+    struct Node *n = NULL;
+    enQueue(self);
+
+    while((n = deQueue()) != NULL) {
+
+        if(n->data==value) {
+            return n;
+        }
+        if(n->l_node) {
+            enQueue(n->l_node);
+        }
+        if(n->r_node) {
+            enQueue(n->r_node);
+        }
+    }
+    return n;
+}
+
+struct Node* tree_dfs_search(struct Node* self, int value) {
+    /*
+    base case: value is found OR not
+    rec. step:
+        1. explore current node
+            if found: return value
+        2. if value is not in current node
+            a. recursive exploration of left subtree
+                if found: return value
+        3. if value is not in left subtree
+            a. recursive exploration of right subtree
+                if found: return value
+
+        if not found: return NULL
+    final state:
+        1. value is returned. algorithm is stopped.
+        2. NULL is returned. all nodes are visited
+    */
+
+   struct Node* n = NULL;
+
+    if(self->data == value) {
+        n = self;
+        return n;
+    }
+    if(self->l_node) {
+        n = tree_dfs_search(self->l_node, value);
+        if((n) && n->data == value) {
+            return n;
+        }
+    }
+    if(self->r_node) {
+        n = tree_dfs_search(self->r_node, value);
+        if((n) && n->data == value) {
+            return n;
+        }
+    }
+
+    return NULL;
+}
+
+void tree_postorder_traverse(struct Node* self) {
+    if (self->l_node) {
+        tree_postorder_traverse(self->l_node);
+    }
+    if (self->r_node) {
+        tree_postorder_traverse(self->r_node);
+    }
+
+    printf("%d\n", self->data);
+}
+
+void tree_preorder_traverse(struct Node* self) {
+    printf("%d\n", self->data);
+
+    if (self->l_node) {
+        tree_preorder_traverse(self->l_node);
+    }
+
+    if (self->r_node) {
+        tree_preorder_traverse(self->r_node);
+    }
+}
+
+void tree_inorder_traverse(struct Node* self) {
+    /*
+    base case:
+        all nodes in subtree are visited
+    rec. step:
+        if l_node
+            rec(l_node)
+        else
+            print data
+        print data
+        if r_node
+            rec(r_node)
+        else
+            print data
+    final state:
+        all nodes are visited
+    */
+
+    if (self->l_node) {
+        tree_inorder_traverse(self->l_node);
+    }
+
+    printf("%d\n", self->data);
+
+    if (self->r_node) {
+        tree_inorder_traverse(self->r_node);
+    }
+}
+
+void tree_delete_node(Tree* tree, struct Node* self) {
+    //if duplicate is encountered: delete first found node
```

Comment:
duplicates can't exist in bin tree: there can be only one node with particular value